### PR TITLE
Feature handler limitranger

### PR DESCRIPTION
**Pull request by whwalter**
Makes the LimitRanger and sourcing the limit range config a top level handler concern and passes the needed limit range context to the child handlers as part of the context since Handle is an interface signature.

### CODE REVIEW

File: internal/admission/admission.go [hunk 1]
```diff
@@ -10,6 +10,10 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
+type LimitRangeContextType string
+
+const LimitRangeContextTypeMemory = "memory"
```
Comment (dskatz) [inline internal/admission/admission.go]:
```suggestion
const LimitRangeContextTypeMemory LimitRangeContextType  = "memory"
```

File: internal/admission/admission.go [hunk 2]
```diff
@@ -65,9 +71,20 @@ func (r *Router) InjectDecoder(d *admission.Decoder) error {
...
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("resource %s not implemented", req.RequestKind.Kind))
+	}
+
+	cfg, err := r.limitRanger.LimitRangeConfig(req.Namespace)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to retrieve limit range information from namespace %s: %s", req.Namespace, err.Error()))
+	}
+
+	if cfg == nil {
+		return admission.Allowed(fmt.Sprintf("No container limit range in namespace: %s", req.Namespace))
```
Comment (dskatz) [inline internal/admission/admission.go:85]:
```suggestion
		return admission.Allowed(fmt.Sprintf("no container limit range in namespace: %s", req.Namespace))
```

File: internal/admission/admission.go [hunk 3]
```diff
@@ -31,12 +35,14 @@ func WithAdmissionHandlers(handlers ...AdmissionHandler) OptionsFunc {
 }
 
 type Router struct {
-	handlers map[string]AdmissionHandler
+	handlers    map[string]AdmissionHandler
+	limitRanger LimitRanger
 }
 
-func NewRouter(opts ...OptionsFunc) (*Router, error) {
+func NewRouter(lr LimitRanger, opts ...OptionsFunc) (*Router, error) {
```
Comment (dskatz) [inline internal/admission/admission.go:42]: The signature change is causing an error in the `cli` package.

```
admissionRouter, err := admission.NewRouter(lr)
```

Comment (whwalter) [reply internal/admission/admission.go:42]: fixed.

**Review by dskatz — APPROVED** [review]

File: internal/admission/admission.go [hunk 4]
```diff
@@ -65,9 +71,20 @@ func (r *Router) InjectDecoder(d *admission.Decoder) error {
...
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to retrieve limit range information from namespace %s: %s", req.Namespace, err.Error()))
+	}
+
+	if cfg == nil {
+		return admission.Allowed(fmt.Sprintf("No container limit range in namespace: %s", req.Namespace))
 	}
 
-	return admission.Errored(http.StatusBadRequest, fmt.Errorf("resource %s not implemented", req.RequestKind.Kind))
+	ctx = context.WithValue(ctx, LimitRangeContextTypeMemory, cfg)
```
Comment (yuzhouliu9) [inline internal/admission/admission.go:88]: Sorry late to the party but the only downside using context.WithValue is that each Handler needs to do dynamic type checking:

```
	config, ok := ctx.Value(admission.LimitRangeContextTypeMemory).(*limitrange.Config)
	if !ok {
		err := fmt.Errorf("failed to get Config from context")
		klog.Log.Error(err, "context does not contain %s", admission.LimitRangeContextTypeMemory)
		return admissionruntime.Errored(http.StatusInternalServerError, err)
	}
```

No preference from me but thought I'd bring it up

**Pull request merged**